feat: add schema aware stream writer

arturowczarek opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [X] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigquerystorage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [X] Ensure the tests and linter pass
- [X] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #2039 ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

arturowczarek committed:
feat: add schema aware stream writer

arturowczarek commented:
This is a work in progress. Please let me know what you think.

Neenu1995 commented:
@yirutang PTAL

arturowczarek commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/SchemaAwareStreamWriter.java:
```diff
@@ -0,0 +1,651 @@
+ * order of minutes).
+ */
+public class SchemaAwareStreamWriter<T> implements AutoCloseable {
+  private static String streamPatternString =
```
Fields like this should be final and CAPITALIZED, but I leave it this way. Otherwise I'd have to change too much.

arturowczarek commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/SchemaAwareStreamWriter.java:
```diff
@@ -0,0 +1,651 @@
+ * table schema is updated, users will be able to ingest data on the new schema after some time (in
+ * order of minutes).
+ */
+public class SchemaAwareStreamWriter<T> implements AutoCloseable {
```
This class is intended to be a generic version of JsonStreamWriter

arturowczarek commented on google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/SchemaAwareStreamWriter.java:
```diff
@@ -0,0 +1,651 @@
+  private Descriptor descriptor;
+  private TableSchema tableSchema;
+  private boolean ignoreUnknownFields = false;
+  private boolean reconnectAfter10M = false;
```
It doesn't seem to be necessary. Can we remove it?

arturowczarek commented:
@yirutang @Neenu1995 Is there a chance for any feedback to this pull request and/or the created issues?

yirutang reviewed (approved):
Thanks for the addition!

arturowczarek commented:
I think I've changed everything. The tests for `JsonStreamWriter` all pass. You can run the CI tests and do the review.

arturowczarek commented:
Thank you @yirutang for running the tests again.
I've fixed the formatting. Sorry about that. I didn't notice `kokoro` requires additional environmental variable `JOB_TYPE` to know what to do.
I'm not sure what to do with clirr errors. It notices that I've introduced some changes to public interfaces: https://github.com/googleapis/java-bigquerystorage/actions/runs/4606032456/jobs/8148652677#step:5:3513
This includes renaming one of the class, removal of `reconnectAfter10M`, etc. I'm not sure what is the workflow and how do deal with it properly. Can you instruct me what to do? I've noticed @Neenu1995 commit https://github.com/googleapis/java-bigquerystorage/commit/e67e913f34fda4f4cc523c0248e5344232c0b736, where she leaves the old, wrongly named class but leaves in the state that doesn't break the contract. When and how is it going to be removed? Should I also leave the classes and methods and mark them somehow that they are going to be removed?

arturowczarek committed:
[squash this commit] Fix clirr errors

arturowczarek commented:
@Neenu1995 Why didn't you finally use `@Deprecated` annotation in https://github.com/googleapis/java-bigquerystorage/pull/2037?

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

arturowczarek commented:
How does releasing of this library work? When does the releasing bot activate and release a new version?
